doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,369 @@
+_`BPO 18835`: https://bugs.python.org/issue18835
+_`PEP 445`: https://www.python.org/dev/peps/pep-0445/
+
+Usage and Impact
+----------------
+
+The new functions can only be accessed via the NumPy C-API. An example is
+included later in the NEP. The added ``struct`` will increase the size of the
+``ndarray`` object. It is a necessary price to be paid for this approach. We
+can be reasonably sure that the change in size will have a minimal impact on
+end-user code because NumPy version 1.20 already changed the object size.
+
+The implementation preserves the use of ``PyTraceMalloc_Track`` to track
+allocations already present in NumPy.
+
+Backward compatibility
+----------------------
+
+The design will not break backward compatibility. Projects that were assigning
+to the ``ndarray->data`` pointer were already breaking the current memory
+management strategy (backed by ``npy_alloc_cache``) and should restore
+``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
+in size should not impact end-users.
+
+Detailed description
+--------------------
+
+High level design
+=================
+
+Users who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
+on the data ``ptr``.
+
+NumPy C-API functions
+=====================
+
+.. c:type:: PyDataMem_Handler
+
+    A struct to hold function pointers used to manipulate memory
+
+    .. code-block:: c
+
+        typedef struct {
+            char name[128];  /* multiple of 64 to keep the struct unaligned */
+            PyDataMem_AllocFunc *alloc;
+            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
+            PyDataMem_FreeFunc *free;
+            PyDataMem_ReallocFunc *realloc;
+        } PyDataMem_Handler;
+
+    where the function's signatures are
+
+    .. code-block:: c
+
+        typedef void *(PyDataMem_AllocFunc)(size_t size);
+        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
+        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
+        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);
+        typedef void *(PyDataMem_CopyFunc)(void *dst, const void *src, size_t size);
```

Comment:
Is `PyDataMem_CopyFunc` a leftover from the previous edit? I don't find it in the struct.

Reply:
removing